### ISSUE TITLE
SO_REUSEADDR in cifsd_tcp_reuseaddr not working?

### ISSUE THREAD

**Issue by Andy2244** [cifsd-team/cifsd#153]
I get this error on a restart sequence, if i have active smb connections (explorer).
```
user.notice cifsd: Stopping CIFS/SMB userspace service.
local5.err cifsd: [cifsd-worker/7362]: ERROR: Child received signal: 15 (Terminated)
local5.info cifsd: [cifsd-manager/7361]: INFO: Exiting. Bye!
daemon.err rmmod: unloading the module failed
daemon.err modprobe: cifsd is already loaded
user.notice cifsd: Starting CIFS/SMB userspace service.
local5.err cifsd: [cifsd-worker/7581]: ERROR: Can't open `/etc/cifs/cifsdpwd.db': No such file or directory
local5.err cifsd: [cifsd-worker/7581]: ERROR: User database file does not exist. Only guest sessions (if permitted) will work.
kern.err kernel: [  591.033915] kcifsd: handle_startup_event:343: Reconnect to a new user space daemon
kern.err kernel: [  591.042202] kcifsd: create_socket:410: Failed to bind socket: -98
kern.err kernel: [  591.048448] kcifsd: tcp_destroy_socket:368: Failed to shutdown socket: -107
kern.err kernel: [  591.055547] Failed to init TCP subsystem: -98
kern.err kernel: [  591.060633] kcifsd: stop_sessions:374: Stop session handler kcifsd:4181/7370
```
That the module can't be unloaded is expected, since we use `ipc timeout = 8` and the restart sequence does not wait 8s.

The problem seems that the listening connection is not closed correctly, since we still have this connection lingering around, even after killing all cifsd processes and rmmod cifsd.
```
netstat -lanp | grep 445
tcp        0      0 0.0.0.0:445             0.0.0.0:*               LISTEN      -
```
From this point on, cifsd will fail with the above errors? Is there some check missing that closes this connection in either the tools or module?

Also why is `SO_REUSEADDR` not working in this case? The connection is already there, so should this not allow to reuse it and prevent the `bind -98` error?

PS: The init stop() is just this for testing
```
killall cifsd
sleep 2
rmmod cifsd
```

start() is this:
```
modprobe cifsd
/usr/sbin/cifsd
```

**Comment by namjaejeon** [comment]
Yup, reproduced it. checking it.
Thanks for your reporting!

**Comment by namjaejeon** [comment]
We can simply reproduce it.
insmod cifsd.ko
./cifsd

killall cifsd
./cifsd
kcifsd: create_socket:410: Failed to bind socket: -98
....
And it doesn't recover even thought rmmod cifsd and restart cifsd again.

**Comment by namjaejeon** [comment]
I pushed the patches to master. please check it is improved.
Thanks!

**Comment by neheb** [comment]
Doesn't seem to have worked:

```
root@OpenWrt:~# logread | grep cif
Wed Jul 17 12:24:16 2019 user.notice cifsd: Starting CIFS/SMB userspace service.
root@OpenWrt:~# /etc/init.d/cifsd stop
root@OpenWrt:~# logread | grep cif
Wed Jul 17 12:24:16 2019 user.notice cifsd: Starting CIFS/SMB userspace service.
root@OpenWrt:~# /etc/init.d/cifsd start
root@OpenWrt:~# logread | grep cifs
Wed Jul 17 12:24:16 2019 user.notice cifsd: Starting CIFS/SMB userspace service.
Wed Jul 17 13:08:47 2019 user.notice cifsd: Starting CIFS/SMB userspace service.
Wed Jul 17 13:08:47 2019 kern.err kernel: [  213.984999] kcifsd: create_socket:418: Failed to bind socket: -125
Wed Jul 17 13:08:47 2019 kern.err kernel: [  213.997390] kcifsd: tcp_destroy_socket:376: Failed to shutdown socket: -134
root@OpenWrt:~# 
```

**Comment by namjaejeon** [comment]
Oh.. It seems another issue. because we are disscussing "cifsd restart" working.

I have a question to reproduce it on my side.

> root@OpenWrt:~# /etc/init.d/cifsd stop

What sequence does it work ?
e.g.
killall cifsd
rmmod cifsd ?

**Comment by neheb** [comment]
The stop and start is found here: https://github.com/openwrt/packages/blob/master/net/cifsd-tools/files/cifsd.init#L141

It does an rmmod before modprobe.

I tested this with /etc/init.d/cifsd stop and stat after ~30 seconds. ps shows this:

```
root@OpenWrt:~# ps | grep cifsd
 2136 root      2660 S    /usr/sbin/cifsd --n
 2138 root      2688 S    /usr/sbin/cifsd --n
 2148 root      1240 S    grep cifsd
root@OpenWrt:~# /etc/init.d/cifsd stop
root@OpenWrt:~# ps | grep cifsd
 2163 root      1240 S    grep cifsd
```

The module is still there. Issuing rmmod before launching it again does nothing. I also see this:

```
root@OpenWrt:~# netstat -an
Active Internet connections (servers and established)
Proto Recv-Q Send-Q Local Address           Foreign Address         State        
tcp        0      0 0.0.0.0:445             0.0.0.0:*               LISTEN      
```

since the processes are dead, I have no idea what's controlling the socket.

I tried with ss -lna. It's not giving me the owner of the socket. Maybe killing cifsd is not correct?

**Comment by namjaejeon** [comment]
I made a mistake in previous patch. I fix it and push it to master now. please check it!
Thanks for your report!

**Comment by neheb** [comment]
Fixed. As for @Andy2244 's issue, I can't replicate on Linux (Windows is not booting anymore for some reason).

**Comment by Andy2244** [comment]
ok tested and restart + zombie connection is fixed

**Comment by Andy2244** [comment]
ok just had a case where i stopped the service and than directly started the samba service, which resulted in a failed startup of samba, because it could not open its listing connection on localhost 445.

So we really need a clean and direct way to shutdown/unload cifsd completely, without those `ipc timeout` delays.

Can you add some kernel module or tools options that forces a full close + unload, without any timeouts?

**Comment by Andy2244** [comment]
Also managed to get:
```
kcifsd: create_socket:418: Failed to bind socket: -98
kcifsd: tcp_destroy_socket:376: Failed to shutdown socket: -107
Failed to init TCP subsystem: -98
```
I think i did a rmmod, while the tools where still running, than killed the tools and ended up with a zombie connection again.

**Comment by sergey-senozhatsky** [comment]
On (07/21/19 06:00), Andy Walsh wrote:
>    ok just had a case where i stopped the service and than directly started
>    the samba service, which resulted in a failed startup of samba, because it
>    could not open its listing connection on localhost 445.
>
>    So we really need a clean and direct way to shutdown/unload cifsd
>    completely, without those ipc timeout delays.

Hmm, OK. That's a bit strange reason to add that functionality, to be
honest, but let's re-open the discussion.

>    Can you add some kernel module or tools options that forces a full close +
>    unload, without any timeouts?

Hmm, sysfs knob or module param, which would require root?

	-ss

**Comment by Andy2244** [comment]
> Hmm, OK. That's a bit strange reason to add that functionality, to be honest, but let's re-open the discussion.

Keep in mind we already use a reduced `ipc timeout`, so it would take much longer to actually stop cifsd service with default settings and being able to start any samba version or fully restart cifsd. I think there should be a way to full stop/unload any service without lengthy extra timeouts.

> Hmm, sysfs knob or module param, which would require root?

Sure, you need root anyway to startup the service and load the module.

**Comment by namjaejeon** [comment]
> Keep in mind we already use a reduced ipc timeout, so it would take much longer to actually stop cifsd service with default settings and being able to start any samba version or fully restart cifsd. I think there should be a way to full stop/unload any service without lengthy extra timeouts.

Yep, understood. Let me check it.
Thanks for your opinion.

**Comment by Andy2244** [comment]
@sergey-senozhatsky So going by your latest commit, this is now a working way to shutdown the cifsd service?
`echo hard > /sys/class/cifsd-control/kill_server`

**Comment by sergey-senozhatsky** [comment]
On (08/07/19 01:40), Andy Walsh wrote:
>    [1]@sergey-senozhatsky So going by your latest commit, this is now a
>    working way to shutdown the cifsd service?
>    echo hard > /sys/class/cifsd-control/kill_server

Yes, but I deliberately didn't mention it. I'd say that this functionality
may be reworked. We still haven't discussed it with Namjae properly.

	-ss

**Comment by Andy2244** [comment]
@sergey-senozhatsky I did retest the latest master and it seems i still need a 4-5s delay between the killswitch and the rmmod, otherwise the rmmod fails, if i have a open smb connection?

```
stop_service()
{
	logger -t 'cifsd' "Stopping CIFSD userspace service."
	killall cifsd > /dev/null 2>&1
	sleep 1
	[ -e /sys/class/cifsd-control/kill_server ] && echo hard > /sys/class/cifsd-control/kill_server
	sleep 5
	[ -e /sys/module/cifsd ] && rmmod cifsd
	[ -f /tmp/cifsd.lock ] && rm /tmp/cifsd.lock
}
```

PS: I also see this in the logs:
```
Stopping CIFSD userspace service.
kern.err kernel: [779344.762364] kcifsd: kill_server_store:434: kill command received
kern.err kernel: [779344.768604] kcifsd: stop_sessions:373: Stop session handler kcifsd:10758/14394
kern.err kernel: [779347.358496] kcifsd: cifsd_conn_handler_loop:298: sock_read failed: -108
```

**Comment by sergey-senozhatsky** [comment]
On (08/20/19 06:51), Andy Walsh wrote:
>    [1]@sergey-senozhatsky I did retest the latest master and it seems i still
>    need a 4-5s delay between the killswitch and the rmmod, otherwise the
>    rmmod fails, if i have a open smb connection?
> 
>  stop_service()
>  {
>          logger -t 'cifsd' "Stopping CIFSD userspace service."
>          killall cifsd > /dev/null 2>&1
>          sleep 1
>          [ -e /sys/class/cifsd-control/kill_server ] && echo hard > /sys/class/cifsd-control/kill_server
>          sleep 5
>          [ -e /sys/module/cifsd ] && rmmod cifsd
>          [ -f /tmp/cifsd.lock ] && rm /tmp/cifsd.lock
>  }
> 
>    PS: I also see this in the logs:

Hmm...
Need to look at this more. This "5 seconds" timeout sounds a bit
suspicious. I wonder if this is a sk->sk_sndtimeo (which is 5 seconds)
that we are having problems with.

The fundamental problem is that we cannot have a really fast killswitch,
which would work all the time. The fact is, we even cannot have a
killswitch, which would simply work all the time. By design. What we
have is

 master kthreads                (per interface)
 	TCP connection kthread  (per connection)
 		worker kthread  (per SMB request (including compound requests))

We ask the server to reset/kill itself we set a global variable, stop
master kthreads (per interface) and then attempt to stop existing TCP
connection kthreads (per connection).

And this is where we have the problem. The TCP connection kthreads and
worker kthreads can block (including UNINTERRUPTIBLE sleep, I suppose),
they can deadlock, they can crash (a NULL pointer deref or an invalid
pagefault, etc.) and so on. Which means that we cannot easily kill them.
To make it even worse, we don't even know what to kill. Unlike master
kthreads or TCP connection kthreads, we don't have task_struct pointers
to worker kthreads, because cifsd uses system-wide work queue; so those
are system (general purpose) workers that do the actual job. Should any
of them block (or crash) we may not be able to kill the server at all.

We can, probably, improve some things here. But bad scenarios will
remain bad for cifsd.

	-ss

**Comment by Andy2244** [comment]
ok thanks for the info, i hope we can improve this. In my tests most of the time it worked with the extra timeout, at around 8-9 seconds it seem to always work. If no connections are open only 1-2 seconds seem needed.

**Comment by sergey-senozhatsky** [comment]
On (08/21/19 03:33), Andy Walsh wrote:
>    ok thanks for the info, i hope we can improve this. In my tests most of
>    the time it worked with the extra timeout, at around 8-9 seconds it seem
>    to always work. If no connections are open only 1-2 seconds seem needed.

Off chance, would anything change with this patch?
Just curious.

kernel TCP recv/snd loops are designed to be used from the user space,
mostly.

---
diff --git a/transport_tcp.c b/transport_tcp.c
index b72dcba..387a50e 100644
--- a/transport_tcp.c
+++ b/transport_tcp.c
@@ -232,8 +232,8 @@ static int cifsd_kthread_fn(void *p)
 		}
 
 		cifsd_debug("connect success: accepted new connection\n");
-		client_sk->sk->sk_rcvtimeo = CIFSD_TCP_RECV_TIMEOUT;
-		client_sk->sk->sk_sndtimeo = CIFSD_TCP_SEND_TIMEOUT;
+		client_sk->sk->sk_rcvtimeo = 1 * HZ;
+		client_sk->sk->sk_sndtimeo = 1 * HZ;
 
 		cifsd_tcp_new_connection(client_sk);
 	}

**Comment by Andy2244** [comment]
@sergey-senozhatsky ok sorry for the late reply.
I still get some weird restart bugs and tried your patch, yet the lowest time i can get is 4-5s

I see this for a 2s extra timeout, with open connections.
```
15:30:40 2019 user.notice cifsd: Stopping CIFSD userspace service.
15:30:40 2019 user.notice cifsd: rmmod 1
15:30:40 2019 daemon.err rmmod: unloading the module failed
15:30:40 2019 user.notice cifsd: killswitch
15:30:40 2019 kern.err kernel: [95843.369445] kcifsd: kill_server_store:436: kill command received
15:30:40 2019 kern.err kernel: [95843.375658] kcifsd: stop_sessions:404: Stop session handler kcifsd:52337/13679
15:30:42 2019 user.notice cifsd: rmmod 2
15:30:42 2019 daemon.err rmmod: unloading the module failed
15:30:42 2019 user.notice cifsd: module still loaded after 2s timeout
```
The stop logic is this:
stop_service()
{
	logger -t 'cifsd' "Stopping CIFSD userspace service."
	killall cifsd > /dev/null 2>&1
	logger -t 'cifsd' "rmmod 1"
	[ -e /sys/module/cifsd ] && rmmod cifsd > /dev/null 2>&1
	# With open smb connections rmmod takes longer
	if [ -e /sys/module/cifsd ]; then
		logger -t 'cifsd' "killswitch"
		[ -e /sys/class/cifsd-control/kill_server ] && echo hard > /sys/class/cifsd-control/kill_server
		sleep 2
		logger -t 'cifsd' "rmmod 2"
		rmmod cifsd > /dev/null 2>&1
	fi
	if [ -e /sys/module/cifsd ]; then
		logger -t 'cifsd' "module still loaded after 2s timeout"
	fi
	[ -f /tmp/cifsd.lock ] && rm /tmp/cifsd.lock
}

**Comment by namjaejeon** [comment]
@Andy2244  Could your script check whether server start-up  from sysfs state ?
cat  /sys/class/cifsd-control/stats
2 startup 445 50

So I suggest the following sequences.
 - killall cifsd
 - echo hard > /sys/class/cifsd-control/kill_server
 - polling if sysfs stats is changed to "startup"
 - if startup, rmmod cifsd

kill_server is trying to reset server asynchorously using workqueue. So it is wrong way that find server reset complete using sleep.

**Comment by Andy2244** [comment]
Sure i can do this, still feels convoluted just to stop a service cleanly?

**Comment by Andy2244** [comment]
@namjaejeon ok as reference, here is my new kill_server code, which works:
```
kill_server()
{
	if [ -e /sys/module/cifsd ]; then
		if [ -e /sys/class/cifsd-control/kill_server ]; then
			logger -p daemon.info -t 'cifsd' "triggering kill_server"
			echo hard > /sys/class/cifsd-control/kill_server
			sleep 1
			local timeout=0
			# we need a extra timeout for the reset (cifsd-control: reset startup running)
			while [ -n "`cat /sys/class/cifsd-control/stats | grep reset`" ]; do
				if [ $timeout -ge 10 ]; then
					return
				fi
				timeout=`expr $TIMEOUT + 1`
				sleep 1
			done
		fi
	fi
}
```

**Comment by Andy2244** [comment]
@namjaejeon ok i tested your email patch and seems to work, except for some minor log issue.

What i observe is that `echo hard > /sys/class/cifsd-control/kill_server` does not directly return anymore, but still needs 2-4 seconds to finish, yet i don't have to poll anymore or add extra timeouts.

Yet we need some log cleanup please, here is what we get after a kill_server:
```
Wed Dec 11 16:21:00 2019 kern.err kernel: [ 1036.172089] kcifsd: kill_server_store:437: kill command received
Wed Dec 11 16:21:00 2019 kern.err kernel: [ 1036.178325] kcifsd: stop_sessions:414: Stop session handler kcifsd:34893/6081
Wed Dec 11 16:21:00 2019 kern.err kernel: [ 1036.297631] kcifsd: stop_sessions:414: Stop session handler kcifsd:34893/6081
Wed Dec 11 16:21:00 2019 kern.err kernel: [ 1036.407728] kcifsd: stop_sessions:414: Stop session handler kcifsd:34893/6081
Wed Dec 11 16:21:00 2019 kern.err kernel: [ 1036.517828] kcifsd: stop_sessions:414: Stop session handler kcifsd:34893/6081
Wed Dec 11 16:21:00 2019 kern.err kernel: [ 1036.627927] kcifsd: stop_sessions:414: Stop session handler kcifsd:34893/6081
Wed Dec 11 16:21:00 2019 kern.err kernel: [ 1036.738024] kcifsd: stop_sessions:414: Stop session handler kcifsd:34893/6081
Wed Dec 11 16:21:00 2019 kern.err kernel: [ 1036.848134] kcifsd: stop_sessions:414: Stop session handler kcifsd:34893/6081
Wed Dec 11 16:21:00 2019 kern.err kernel: [ 1036.958221] kcifsd: stop_sessions:414: Stop session handler kcifsd:34893/6081
Wed Dec 11 16:21:00 2019 kern.err kernel: [ 1037.068320] kcifsd: stop_sessions:414: Stop session handler kcifsd:34893/6081
Wed Dec 11 16:21:01 2019 kern.err kernel: [ 1037.178418] kcifsd: stop_sessions:414: Stop session handler kcifsd:34893/6081
Wed Dec 11 16:21:01 2019 kern.err kernel: [ 1037.298527] kcifsd: stop_sessions:414: Stop session handler kcifsd:34893/6081
Wed Dec 11 16:21:01 2019 kern.err kernel: [ 1037.418631] kcifsd: stop_sessions:414: Stop session handler kcifsd:34893/6081
Wed Dec 11 16:21:01 2019 kern.err kernel: [ 1037.528729] kcifsd: stop_sessions:414: Stop session handler kcifsd:34893/6081
Wed Dec 11 16:21:01 2019 kern.err kernel: [ 1037.638830] kcifsd: stop_sessions:414: Stop session handler kcifsd:34893/6081
Wed Dec 11 16:21:01 2019 kern.err kernel: [ 1037.748925] kcifsd: stop_sessions:414: Stop session handler kcifsd:34893/6081
Wed Dec 11 16:21:01 2019 kern.err kernel: [ 1037.859022] kcifsd: stop_sessions:414: Stop session handler kcifsd:34893/6081
Wed Dec 11 16:21:01 2019 kern.err kernel: [ 1037.969121] kcifsd: stop_sessions:414: Stop session handler kcifsd:34893/6081
Wed Dec 11 16:21:01 2019 kern.err kernel: [ 1038.079221] kcifsd: stop_sessions:414: Stop session handler kcifsd:34893/6081
Wed Dec 11 16:21:02 2019 kern.err kernel: [ 1038.199326] kcifsd: stop_sessions:414: Stop session handler kcifsd:34893/6081
Wed Dec 11 16:21:02 2019 kern.err kernel: [ 1038.309424] kcifsd: stop_sessions:414: Stop session handler kcifsd:34893/6081
Wed Dec 11 16:21:02 2019 kern.err kernel: [ 1038.419520] kcifsd: stop_sessions:414: Stop session handler kcifsd:34893/6081
Wed Dec 11 16:21:02 2019 kern.err kernel: [ 1038.529629] kcifsd: stop_sessions:414: Stop session handler kcifsd:34893/6081
Wed Dec 11 16:21:02 2019 kern.err kernel: [ 1038.639728] kcifsd: stop_sessions:414: Stop session handler kcifsd:34893/6081
Wed Dec 11 16:21:02 2019 kern.err kernel: [ 1038.749822] kcifsd: stop_sessions:414: Stop session handler kcifsd:34893/6081
Wed Dec 11 16:21:02 2019 kern.err kernel: [ 1038.859908] kcifsd: stop_sessions:414: Stop session handler kcifsd:34893/6081
Wed Dec 11 16:21:02 2019 kern.err kernel: [ 1038.970006] kcifsd: stop_sessions:414: Stop session handler kcifsd:34893/6081
Wed Dec 11 16:21:02 2019 kern.err kernel: [ 1039.080104] kcifsd: stop_sessions:414: Stop session handler kcifsd:34893/6081
Wed Dec 11 16:21:03 2019 kern.err kernel: [ 1039.190200] kcifsd: stop_sessions:414: Stop session handler kcifsd:34893/6081
Wed Dec 11 16:21:03 2019 kern.err kernel: [ 1039.300300] kcifsd: stop_sessions:414: Stop session handler kcifsd:34893/6081
Wed Dec 11 16:21:03 2019 kern.err kernel: [ 1039.410392] kcifsd: stop_sessions:414: Stop session handler kcifsd:34893/6081
Wed Dec 11 16:21:03 2019 kern.err kernel: [ 1039.520587] kcifsd: stop_sessions:414: Stop session handler kcifsd:34893/6081
```
I had a case where i got like 50+ entries of this.

**Comment by namjaejeon** [comment]
okay, Could you test one more after changing socket timeout to 1sec from 7sec and 5sec like this ?

connection.h
#define CIFSD_TCP_RECV_TIMEOUT (1 * HZ)
#define CIFSD_TCP_SEND_TIMEOUT (1 * HZ)

Thanks!

**Comment by Andy2244** [comment]
@namjaejeon 
ok the kill now takes max. 1sec with open connections. So i guess its safe to use those 2 changes for my next version?

```
Thu Dec 12 17:11:36 2019 daemon.notice cifsd: Starting CIFS/SMB userspace service.
Thu Dec 12 17:11:47 2019 kern.err kernel: [90571.188704] kcifsd: kill_server_store:437: kill command received
Thu Dec 12 17:11:47 2019 kern.err kernel: [90571.194914] kcifsd: stop_sessions:414: Stop session handler kcifsd:3981/10097
Thu Dec 12 17:11:47 2019 kern.err kernel: [90571.304827] kcifsd: stop_sessions:414: Stop session handler kcifsd:3981/10097
Thu Dec 12 17:11:47 2019 kern.err kernel: [90571.414904] kcifsd: stop_sessions:414: Stop session handler kcifsd:3981/10097
Thu Dec 12 17:11:47 2019 kern.err kernel: [90571.524982] kcifsd: stop_sessions:414: Stop session handler kcifsd:3981/10097
Thu Dec 12 17:11:47 2019 kern.err kernel: [90571.635066] kcifsd: stop_sessions:414: Stop session handler kcifsd:3981/10097
Thu Dec 12 17:11:47 2019 kern.err kernel: [90571.745139] kcifsd: stop_sessions:414: Stop session handler kcifsd:3981/10097
```

**Comment by namjaejeon** [comment]
> ok the kill now takes max. 1sec with open connections. So i guess its safe to use those 2 changes for my next version?

Let me check more. I am checking side effect when changing default timeout value.

> Still can we get rid of those kern.err messages, except for the kill command + maybe just one line for the Stop session handler?

Yes, I will change it with debug message.

> So if we send the kill, we expect those errors and ignore them?

Yes. Right.

**Comment by namjaejeon** [comment]
@Andy2244 I sent the patch to you. This patch set zero to receive and send timeout before socket shutdown. So this patch don't touch default timeout. And actually I can not reproduce it on my PC, So Could you please check one more if this patch work as we expect or not?

Thanks!

**Comment by Andy2244** [comment]
@namjaejeon 
```
/root/openwrt/build_dir/target-arm_cortex-a9+vfpv3_musl_eabi/linux-mvebu_cortexa9/cifsd-2019-12-07-76063c33/transport_tcp.c: In function 'cifsd_tcp_rev_timeout':
/root/openwrt/build_dir/target-arm_cortex-a9+vfpv3_musl_eabi/linux-mvebu_cortexa9/cifsd-2019-12-07-76063c33/transport_tcp.c:59:38: error: 'SO_RCVTIMEO_OLD' undeclared (first use in this function); did you mean 'SO_RCVTIMEO'?
  kernel_setsockopt(sock, SOL_SOCKET, SO_RCVTIMEO_OLD, (char *)&tv,
                                      ^~~~~~~~~~~~~~~
                                      SO_RCVTIMEO
/root/openwrt/build_dir/target-arm_cortex-a9+vfpv3_musl_eabi/linux-mvebu_cortexa9/cifsd-2019-12-07-76063c33/transport_tcp.c:59:38: note: each undeclared identifier is reported only once for each function it appears in
/root/openwrt/build_dir/target-arm_cortex-a9+vfpv3_musl_eabi/linux-mvebu_cortexa9/cifsd-2019-12-07-76063c33/transport_tcp.c: In function 'cifsd_tcp_snd_timeout':
/root/openwrt/build_dir/target-arm_cortex-a9+vfpv3_musl_eabi/linux-mvebu_cortexa9/cifsd-2019-12-07-76063c33/transport_tcp.c:67:38: error: 'SO_SNDTIMEO_OLD' undeclared (first use in this function); did you mean 'SO_SNDTIMEO'?
  kernel_setsockopt(sock, SOL_SOCKET, SO_SNDTIMEO_OLD, (char *)&tv,
                                      ^~~~~~~~~~~~~~~
                                      SO_SNDTIMEO

```

**Comment by namjaejeon** [comment]
@Andy2244  Oops, Sorry. I sent the patch again. please check it.

**Comment by Andy2244** [comment]
@namjaejeon ok new patch works and seems to kill in <=1sec, so closing this, thanks!

```
Sun Dec 15 01:39:44 2019 daemon.notice cifsd: Stopping CIFSD userspace service.
Sun Dec 15 01:39:44 2019 daemon.err rmmod: unloading the module failed
Sun Dec 15 01:39:44 2019 daemon.info cifsd: triggering kill_server
Sun Dec 15 01:39:45 2019 kern.err kernel: [194865.904212] kcifsd: kill_server_store:437: kill command received
```
